Sources/DarkModeCore/DMTraitCollection.m

```diff
@@ -4,17 +4,127 @@
 //
 
 #import "DMTraitCollection.h"
+#import "UIApplication+DarkModeKit.h"
+
+@import ObjectiveC;
+
+@implementation NSObject (DMTraitEnvironment)
+
++ (void)swizzleTraitCollectionDidChangeToDMTraitCollectionDidChange {
+  [self swizzleTraitCollectionDidChangeToDMTraitCollectionDidChangeWithBlock:nil];
+}
+
++ (void)swizzleTraitCollectionDidChangeToDMTraitCollectionDidChangeWithBlock:(void (^)(NSObject *, UITraitCollection *))block API_AVAILABLE(ios(13.0)) {
+  // Only swizzling classes that conforms to both UITraitEnvironment & DMTraitEnvironment
+  if (!class_conformsToProtocol(self, @protocol(UITraitEnvironment)) || !class_conformsToProtocol(self, @protocol(DMTraitEnvironment))) {
+    return;
+  }
+
+  SEL selector = @selector(traitCollectionDidChange:);
+  Method method = class_getInstanceMethod(self, selector);
+
+  if (!method)
+    NSAssert(NO, @"Method not found for [%@ traitCollectionDidChange:]", NSStringFromClass(self));
+
+  IMP imp = method_getImplementation(method);
+  class_replaceMethod(self, selector, imp_implementationWithBlock(^(NSObject *self, UITraitCollection *previousTraitCollection) {
+    // Call previous implementation
+    ((void (*)(NSObject *, SEL, UITraitCollection *))imp)(self, selector, previousTraitCollection);
+
+    // Call DMTraitEnvironment
+    [(id <DMTraitEnvironment>)self dmTraitCollectionDidChange:previousTraitCollection == nil ? nil : [DMTraitCollection traitCollectionWithUITraitCollection:previousTraitCollection]];
+
+    // Call custom block
+    if (block) {
+      block(self, previousTraitCollection);
+    }
+  }), method_getTypeEncoding(method));
+}
+
+@end
 
 @implementation DMTraitCollection
 
-static DMTraitCollection *_currentTraitCollection = nil;
+static DMTraitCollection *_lastManuallySetTraitCollection = nil; // This is set manually in setCurrentTraitCollection:animated
+static __weak UIApplication *_registeredApplication = nil;
 
 + (DMTraitCollection *)currentTraitCollection {
-  return _currentTraitCollection;
+  DMTraitCollection *lastTraitCollection = [self lastManuallySetTraitCollection];
+  if (@available(iOS 13.0, *)) {
+    if (lastTraitCollection.userInterfaceStyle == DMUserInterfaceStyleUnspecified) {
+      // Return the ones used from the system when no specific user interface style is explicitly set
+      return [self currentSystemTraitCollection];
+    }
+  }
+  return lastTraitCollection;
```

Comment:
I think this will break view or view controller level `overrideUserInterfaceStyle` too.

Reply:
neither will this.

Reply:
Oh, I see. Colors works fine because we used `UIColor` instead `DynamicColor` in iOS 13.

But I think access `DMTraitCollection.current.userInterfaceStyle` directly will get surprise value:

```swift
final class UIViewVC: ViewController {
  override func viewDidLoad() {
    super.viewDidLoad()

    view.backgroundColor = UIColor(.dm, light: .white, dark: .black)
    if #available(iOS 13.0, *) {
      overrideUserInterfaceStyle = .dark
    }
  }

  override func dmTraitCollectionDidChange(_ previousTraitCollection: DMTraitCollection?) {
    super.dmTraitCollectionDidChange(previousTraitCollection)
    print(DMTraitCollection.current.userInterfaceStyle.rawValue)
  }
}
```

Reply:
ah yes, the UIKit function UITraitCollection.current.userInterfaceStyle will always return .dark in this case, maybe we should let DMTraitCollection.current return UITraitCollection.current instead? I used UIScreen.main.traitCollection in DMTraitCollection.current just to keep the previous behavior.

Reply:
Personally I don't think `DMTraitCollection` should be a control point, it should act as same as `UITraitCollection`.

So I'd prefer return `UITraitCollection.current` directly in iOS 13. And leave control logic to users or design a better API in  another PR.

Reply:
the tricky part for directly returning UITraitCollection.current is that 
If we call setCurrentTraitCollection with .dark style, which override all windows' overrideUserInterfaceStyle to dark, but UITraitCollection.current will still return .light (if not called within traitCollectionDidChange methods etc) if the system theme is white.

Also setting UITraitCollection.current value is useless in this case, because it will be overridden by the system immediately.

But probably on iOS 13 we just should not use DMTraitCollection.current value to determine what theme we are in. that's also a reason why I introduced lastManuallySetTraitCollection in this PR.

Reply:
now it will just return what UITraitCollection.current

Reply:
I don't think we should provide `DMTraitCollection.current` at all. `UITraitCollection.current` is mainly for low-level C graphics APIs. Its value would change constantly. For example, before `draw(_ rect: CGRect)` of every `UIView` instance, `UITraitCollection.current` would be set to the `traitCollection` property of the current `UIView` instance. It is not supposed to be used directly. 

```swift
class ExampleView: UIView {

    override init(frame: CGRect) {
        super.init(frame: frame)
        overrideUserInterfaceStyle = .dark
    }

    override func draw(_ rect: CGRect) {
        super.draw(rect)
        print(UITraitCollection.current === self.traitCollection) // true
        print(UITraitCollection.current.userInterfaceStyle.rawValue) // 2
        DispatchQueue.main.async {
            // next run loop
            print(UITraitCollection.current === self.traitCollection) // false
            print(UITraitCollection.current.userInterfaceStyle.rawValue) // 1
        }
    }

}
```

If developers need to get the user interface style of the device, they can use `UIScreen.main.traitCollection.userInterfaceStyle`; if they want to get the local overridden value, they can use `traitCollection` property of `UIView`. Both properties are available since iOS 8.

I also agree with @kylinroc that we should control the interface style elsewhere, so I believe it'd best we deprecate `DMTraitCollection.current`. How do you think? @eyvallah @kylinroc

Reply:
I agree that as a developer  I would go directly for UIView or UIViewController's traitCollection rather than the global one. Providing a current in DMTraitCollection seems not really useful here, on iOS 12, it simply returns overrideTraitCollection value, while on iOS 13 we don't actually apply the light or dark color by ourselves. Nobody would really be calling this method, so I think it is okay to mark it as deprecated.

So as to providing a new location for controlling the interface style, currently it is by setting overrideTraitCollection for convenience, we can find a better spot in the future.

But for now, maybe let's just keep the change within scope, both can be done in PR afterwards. I will add a checklist issue tracking it.